### ISSUE TITLE
Loss NAN

### ISSUE THREAD

**Issue by nengwp** [RobinWu218/SimDINO#20]
I used my own data set for simdinov2 training with a loss of nan, but dinov2 did not occur.
32*H100 is used, the total batchsize is 2400, the model is vit_so400m, and the patch size is 14.
About 30,000 iterations are trained.  The log before the crash was normal
logs：

I20250403 20:25:17 90 dinov2 helpers.py:102] Training  [  33730/2403600]  eta: 39 days, 17:53:23  dino_mcr_loss: -1.0109 (-0.8978)  dino_mcr_comp_loss: 1.0000 (0.7637)  dino_mcr_global_comp_loss: 1.0000 (0.8465)  dino_mcr_expa_loss: 0.0109 (0.1341)  ibot_comp_loss: -0.5001 (-0.4620)  total_loss: -1.5110 (-1.3598)  lr: 0.0026 (0.0013)  wd: 0.0402 (0.0401)  mom: 0.9000 (0.9000)  last_layer_lr: 0.0026 (0.0012)  current_batch_size: 78.0000 (78.0000)  losses_reduced: -1.0112 (-0.9752)  total_grad_norm: 0.0002 (0.0794)  time: 1.466541  data: 0.000128  max mem: 73009

**Comment by RobinWu218** [comment]
Hi @nengwp , the batch size is quite large in your case and one thing you can do is to decrease `--eps` (because large batch size together with large `--eps` might lead to numerical instability when inverting covariance matrix in the coding rate computation). Another way to mitigate this is via higher precision for the coding rate part as seen in #19 . Let me know if you still encounter this problem.

**Comment by nengwp** [comment]
Thank you for your reply. I visualized the loss changes and found that before NAN, loss was very unstable. This may be the cause of NAN. Can you help me locate the problem?
![Image](https://github.com/user-attachments/assets/2085cb71-f8d8-4158-95e1-43baee036a4a)

**Comment by RobinWu218** [comment]
Yes, these loss spikes are caused by the bad conditioning of the scaled covariance matrix ($\frac{d}{n\epsilon^2}ZZ^{\top}$) in coding rate computation, which can happen with large batch size and high `--eps` (also check this [tip](https://github.com/RobinWu218/SimDINO#faq--tips-on-training)). Have you tried lowering `--eps`?

**Comment by nengwp** [comment]
Thanks for your reply. I tried to reduce the learning rate (1e-4), increase the momentum (0.995) and reduce the eps (0.005). The training is relatively stable at present.

![Image](https://github.com/user-attachments/assets/1ccd187a-5d5b-427d-a749-05a60b1e58a2)

I found that momentum is a very critical parameter. When 0.95 is set, there is also a small fluctuation of change. When set 0.99, the fluctuation is very small.But for limited time training, I changed the learning rate, momentum, eps at the same time.
In addition, if momentum is 0.992 in dinov2 and 0.9 in simdinov2, is there any other intrinsic motivation (e.g., acceleration training) besides proving the robustness of the parameter? There is no discussion about the momentum setting in simdinov2 in the paper, but the sensitivity of dinov2 to momentum is illustrated (0.9 vs 0.992).

**Comment by nengwp** [comment]
Update, train longer, nan happens again

![Image](https://github.com/user-attachments/assets/c1ca2f23-0daf-4125-882c-a8e32cfb10ae)

**Comment by RobinWu218** [comment]
Hi @nengwp, let me try to clarify and help with your issues:

- momentum: this parameter is indeed quite important from our experience. We also found that lowering it (0.992 -> 0.9) slightly improves SimDINO while the same change doesn't work for the original DINO. Other than this empirical evidence, we don't currently have much guidance on how to choose this parameter. I think a more careful sweep is necessary for optimal performance.
- Training instability: it is possible that you need to lower `--eps` further, as shown in the large spike before the collapse. Also, to eliminate other potential factors, have you tried with using our default [configs](https://github.com/RobinWu218/SimDINO/blob/main/simdinov2/configs/simdino_config.yaml)? 

Let me know if you run into other issues.

**Comment by nengwp** [comment]
When lowering `--eps` to 0.0001, the training directly encounters NAN, which may be a numerical problem caused by the square of the value being too small.
Default configuration training has been used, which makes NAN easier

**Comment by RobinWu218** [comment]
That might have been too small. Empirically, we find the keeping eps inversely proportional to bs to be sufficient. Additionally, I am not sure what the momentum is set here (can't really tell from the plot). 

We have never encountered NaN when using the default configs. Did you change other parameters (e.g. patch size)?

**Comment by nengwp** [comment]
momenum set 0.995 or 0.999。 I set up a larger network, vit_so400m, and used a patch size of 14.

![Image](https://github.com/user-attachments/assets/ff90d887-f2d1-4d89-abbf-4acdbfd57a3a)

[config_yaml.txt](https://github.com/user-attachments/files/19806795/config_yaml.txt)

When smaller momenta, such as 0.9, are used, NAN comes earlier, and so does a high learning rate.

Although setting a larger momentum and a smaller learning rate can delay the occurrence of NAN, it will still occur as the number of iterations increases. The same configuration file as the log has been given above. Additionally, before the emergence of NAN, the mcr loss showed an upward trend, which seemed inappropriate, especially with a large numerical upward trend. Because a slight upward trend has been observed in your log file. Is this upward trend reasonable? Does this weaken the marker of the loss response training process?

**Comment by RobinWu218** [comment]
Sorry for my late response @nengwp . We always observe a slight upward trend, but never saw anything as drastic as what you observed here. Additionally, the config file that you are using seems quite different to our default [config](https://github.com/RobinWu218/SimDINO/blob/main/simdinov2/configs/simdino_config.yaml) (e.g. projector dimension, epochs, batch_size_per_gpu). Do you have a log when you apply our default config to your custom dataset? It is quite hard to tell what might be the main culprit due to these differences in configurations.

**Comment by nengwp** [comment]
![Image](https://github.com/user-attachments/assets/cbc20224-193a-453f-acdc-794abb8aadd5)

[config_yaml.txt](https://github.com/user-attachments/files/19861581/config_yaml.txt)

Exactly the same experiment was not performed, and I have not done any further experiments now. The closest experiment was to use a larger network with no other hyperparameters changed.

**Comment by RobinWu218** [comment]
Another possibility is that the `batch_size_per_gpu` is small in your setting. It might be worth setting`--reduce_cov=1`. The default setting (i.e. `--reduce_cov=0`) means we do not collect covariance matrices from all the gpus. This can be a problem in your previous run when your projector dim is high (384).

**Comment by nengwp** [comment]
I have 32 H100 images, and the total batchsize is about 2496. Should the batchsize of a single image or all batchsizes be used in the calculation?

**Comment by nengwp** [comment]
Thank you for your answer. I will try your solution if I have time.
